File: backend/functions/proxy.js
/**
 * Proxy Lambda Function
 * Handles app proxy requests from Shopify storefront
 * Can be used to fetch data for custom storefronts
 * Supports multiple app registrations via dynamic credential loading
 */

const crypto = require('crypto');
const { getShopAccessToken, getAppCredentials } = require('./credentials');

/**
 * Verify proxy signature from Shopify
 */
async function verifyProxySignature(queryParams, signature, appId) {
  // Load app credentials to get the secret
  const appCredentials = await getAppCredentials(appId);

  // Remove signature from params
  const params = { ...queryParams };
  delete params.signature;

  // Sort and build query string
  const sortedParams = Object.keys(params)
    .sort()
    .map(key => `${key}=${params[key]}`)
    .join('');

  // Generate signature
  const hash = crypto
    .createHmac('sha256', appCredentials.clientSecret)
    .update(sortedParams)
    .digest('hex');

  return hash === signature;
}

/**
 * Make API call to Shopify
 */
async function callShopifyApi(shop, accessToken, endpoint, method = 'GET', body = null) {
  const url = `https://${shop}/admin/api/2024-01/${endpoint}`;

  const options = {
    method,
    headers: {
      'X-Shopify-Access-Token': accessToken,
      'Content-Type': 'application/json',
    },
  };

  if (body) {
    options.body = JSON.stringify(body);
  }

  const response = await fetch(url, options);

  if (!response.ok) {
    throw new Error(`Shopify API error: ${response.statusText}`);
  }

  return await response.json();
}

exports.handler = async (event) => {
  try {
    const params = event.queryStringParameters || {};
    const { shop, app, signature } = params;

    if (!shop || !app || !signature) {
      return {
        statusCode: 400,
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ error: 'Missing required parameters (shop, app, signature)' }),
      };
    }

    // Verify the signature with the correct app secret
    const isValid = await verifyProxySignature(params, signature, app);
    if (!isValid) {
      return {
        statusCode: 403,
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ error: 'Invalid signature' }),
      };
    }

    // Get access token from Parameter Store
    const accessToken = await getShopAccessToken(app, shop);

    // Parse request body if present
    const requestBody = event.body ? JSON.parse(event.body) : {};
    const { endpoint = 'products.json', method = 'GET' } = requestBody;

    // Call Shopify API
    const data = await callShopifyApi(shop, accessToken, endpoint, method);

    return {
      statusCode: 200,
      headers: {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*',
      },
      body: JSON.stringify(data),
    };
  } catch (error) {
    console.error('Proxy error:', error);
    return {
      statusCode: 500,
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        error: 'Internal server error',
        message: error.message
      }),
    };
  }
};

File: backend/functions/proxy.py
"""
Proxy Lambda Function
Handles app proxy requests from Shopify storefront
Can be used to fetch data for custom storefronts
"""

import os
import json
import hmac
import hashlib

import boto3
import requests

ssm_client = boto3.client('ssm', region_name=os.environ.get('AWS_REGION', 'eu-central-1'))


def get_access_token(shop: str) -> str:
    """Get access token from Parameter Store"""
    prefix = os.environ.get('PARAMETER_STORE_PREFIX', '/shopify/clients')

    response = ssm_client.get_parameter(
        Name=f'{prefix}/{shop}/access-token',
        WithDecryption=True
    )

    return response['Parameter']['Value']


def verify_proxy_signature(query_params: dict, signature: str) -> bool:
    """Verify proxy signature from Shopify"""
    api_secret = os.environ.get('SHOPIFY_API_SECRET')

    # Remove signature from params
    params = {k: v for k, v in query_params.items() if k != 'signature'}

    # Sort and build query string (no separators for proxy signature)
    sorted_params = ''.join(f'{k}={v}' for k, v in sorted(params.items()))

    # Generate signature
    computed_signature = hmac.new(
        api_secret.encode('utf-8'),
        sorted_params.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()

    return hmac.compare_digest(computed_signature, signature)


def call_shopify_api(shop: str, access_token: str, endpoint: str, method: str = 'GET', body: dict = None) -> dict:
    """Make API call to Shopify"""
    url = f'https://{shop}/admin/api/2024-01/{endpoint}'

    headers = {
        'X-Shopify-Access-Token': access_token,
        'Content-Type': 'application/json'
    }

    response = requests.request(
        method=method,
        url=url,
        headers=headers,
        json=body if body else None
    )

    if not response.ok:
        raise Exception(f'Shopify API error: {response.text}')

    return response.json()


def handler(event, context):
    try:
        params = event.get('queryStringParameters') or {}
        shop = params.get('shop')
        signature = params.get('signature')

        if not shop or not signature:
            return {
                'statusCode': 400,
                'headers': {'Content-Type': 'application/json'},
                'body': json.dumps({'error': 'Missing required parameters'})
            }

        # Verify the signature
        if not verify_proxy_signature(params, signature):
            return {
                'statusCode': 403,
                'headers': {'Content-Type': 'application/json'},
                'body': json.dumps({'error': 'Invalid signature'})
            }

        # Get access token from Parameter Store
        access_token = get_access_token(shop)

        # Parse request body if present
        request_body = json.loads(event.get('body', '{}')) if event.get('body') else {}
        endpoint = request_body.get('endpoint', 'products.json')
        method = request_body.get('method', 'GET')

        # Call Shopify API
        data = call_shopify_api(shop, access_token, endpoint, method)

        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*'
            },
            'body': json.dumps(data)
        }
    except Exception as e:
        print(f'Proxy error: {e}')
        return {
            'statusCode': 500,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps({
                'error': 'Internal server error',
                'message': str(e)
            })
        }
